subprojects/docs/src/docs/userguide/feature_variants.adoc

```diff
@@ -0,0 +1,226 @@
+====
+
+[[sec:feature_variant_source_set]]
+=== Feature variant source set
+
+In the previous example, we're declaring a feature variant which uses the _main source set_.
+This is a typical use case in the Java ecosystem, where it's, for whatever reason, not possible to split the sources of a project into different subprojects or different source sets.
+Gradle will therefore declare the configurations as described, but will also setup the compile classpath and runtime classpath of the _main source set_ so that it extends from the feature configuration.
+Said differently, this allows you to declare the dependencies specific to a feature in their own "bucket", but everything is still compiled as a single source set.
+There will also be a single artifact (the component Jar) including support for all features.
+
+However, it is often preferred to have a _separate source set_ for a feature. Gradle will then perform a similar mapping, but will _not_ make the compile and runtime classpath of the main component extend from the dependencies of the registered features.
+It will also, by convention, create a `Jar` task to bundle the classes built from this feature source set, using a classifier corresponding to the kebab-case name of the feature:
+
+.Declaring a feature variant using a separate source set
+====
+include::sample[dir="java-feature-variant/producer-separate-sourceset/groovy",files="build.gradle[tags=register_variant]"]
+include::sample[dir="java-feature-variant/producer-separate-sourceset/kotlin",files="build.gradle.kts[tags=register_variant]"]
+====
+
+[[sec::publishing_feature_variants]]
+== Publishing feature variants
+
+[WARNING]
+====
+Depending on the metadata file format, publishing feature variants may be lossy:
+
+- using {metadata-file-spec}[Gradle metadata], everything is published and consumers will get the full benefit of feature variants
+- using POM metadata (Maven), feature variants are published as **optional dependencies** and artifacts of feature variants are published with different _classifiers_
+- using Ivy metadata, feature variants are lost
+====
+
+Publishing feature variants is supported using the `maven-publish` plugin only.
+The Java Plugin (or Java Library Plugin) will take care of registering the additional variants for you, so there's no additional configuration required, only the regular publications:
+
+.Publishing a component with feature variants
+====
+include::sample[dir="java-feature-variant/producer/groovy",files="build.gradle[tags=plugins,publishing]"]
+include::sample[dir="java-feature-variant/producer/kotlin",files="build.gradle.kts[tags=plugins,publishing]"]
+====
+
+[[sec::consuming_feature_variants]]
+== Dependencies on feature variants
+
+[WARNING]
+====
+As mentioned earlier, feature variants can be lossy when published.
+As a consequence, a consumer can depend on a feature variant only in two cases:
+- with a project dependency (in a multi-project build)
```

Comment:
Missing a blank line causing the list to not render

Reply:
fixed